integrations/snowflake/src/haystack_integrations/components/retrievers/snowflake/snowflake_retriever.py

```diff
@@ -0,0 +1,332 @@
+            if table_name.lower() == privilege[3].lower() and re.match(
+                pattern="truncate|update|insert|delete|operate|references",
+                string=privilege[1],
+                flags=re.IGNORECASE,
+            ):
+                logger.error("User does not have `Select` privilege on the table.")
+                return False
+
+        return True
+
+    def _check_privilege(
+        self,
+        conn: SnowflakeConnection,
+        query: str,
+        user: str,
+    ) -> bool:
+        """
+        Check whether a user has a `select`-only access to the table.
+
+        :param conn: An open connection to Snowflake.
+        :param query: The query from where to extract table names to check read-only access.
+        """
+        cur = conn.cursor()
+
+        cur.execute(f"SHOW GRANTS TO USER {user};")
+
+        # Get user's latest role
+        roles = cur.fetchall()
+        if not roles:
+            logger.error("User does not exist")
+            return False
+
+        # Last row second column from GRANT table
+        role = roles[-1][1]
+
+        # Get role privilege
+        cur.execute(f"SHOW GRANTS TO ROLE {role};")
+
+        # Keep table level privileges
+        table_privileges = [row for row in cur.fetchall() if row[2] == "TABLE"]
+
+        # Get table names to check for privilege
+        table_names = self._extract_table_names(query=query)
+
+        for table_name in table_names:
+            if not self._has_select_privilege(
+                privileges=table_privileges,
+                table_name=table_name,
+            ):
+                return False
+        return True
+
+    def _fetch_data(
+        self,
+        query: str,
+    ) -> pd.DataFrame:
+        """
+        Fetch data from a database using a SQL query.
+
+        :param query: SQL query to use to fetch the data from the database. Query must be a valid SQL query.
+        """
+
+        df = pd.DataFrame()
+        if not query:
+            return df
+        try:
+            # Create a new connection with every run
+            conn = self._snowflake_connector(
+                connect_params={
+                    "user": self.user,
+                    "account": self.account,
+                    "password": self.api_key.resolve_value(),
+                    "database": self.database,
+                    "schema": self.db_schema,
+                    "warehouse": self.warehouse,
+                    "login_timeout": self.login_timeout,
+                }
+            )
+            if conn is None:
+                return df
+        except (ForbiddenError, ProgrammingError) as e:
+            logger.error(
+                "Error connecting to Snowflake ({errno}): {error_msg}",
+                errno=e.errno,
+                error_msg=e.msg,
+            )
+            return df
+
+        try:
+            # Check if user has `select` privilege on the table
+            if self._check_privilege(
```

Comment:
I think it's worth to add an else here and inform the user that he/she don't have enough privilege to read from that table.